service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+package frontend
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/status"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.temporal.io/api/proxy"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/rpc/encryption"
+	"go.temporal.io/server/common/util"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+)
+
+// HTTPAPIServer is an HTTP API server that forwards requests to gRPC via the
+// gRPC interceptors.
+type HTTPAPIServer struct {
+	server                 http.Server
+	listener               net.Listener
+	logger                 log.Logger
+	serveMux               *runtime.ServeMux
+	shutdownDrainTime      time.Duration
+	stopped                chan struct{}
+	matchAdditionalHeaders map[string]bool
+}
+
+var defaultForwardedHeaders = []string{
+	"Authorization-Extras",
+	"X-Forwarded-For",
+	http.CanonicalHeaderKey(headers.ClientNameHeaderName),
+	http.CanonicalHeaderKey(headers.ClientVersionHeaderName),
+}
+
+type httpRemoteAddrContextKey struct{}
+
+// NewHTTPAPIServer creates an [HTTPAPIServer].
+func NewHTTPAPIServer(
+	serviceConfig *Config,
+	rpcConfig config.RPC,
+	grpcListener net.Listener,
+	tlsConfigProvider encryption.TLSConfigProvider,
+	handler Handler,
+	interceptors []grpc.UnaryServerInterceptor,
+	logger log.Logger,
+) (*HTTPAPIServer, error) {
+	// Create a TCP listener the same as the frontend one but with different port
+	tcpAddrRef, _ := grpcListener.Addr().(*net.TCPAddr)
+	if tcpAddrRef == nil {
+		return nil, fmt.Errorf("must use TCP for gRPC listener to support HTTP API")
+	}
```

Comment:
why not just use `ok`?

Reply:
also, I'd include a %T somewhere

Reply:
Don't need `ok`, why use it? I don't think we need Go types on this error message.